# EDA Project

<div class="alert alert-block alert-info">The EDA project in this course has four main parts to it: <br>
    
1. Project Proposal
2. Phase 1
3. Phase 2
4. Report

This notebook will be used for Project Proposal, Phase 1, and Phase 2. You will have specific questions to answer within this notebook for Project Proposal and Phase 1. You will also continue using this notebook for Phase 2. However, guidance and expectations can be found on Canvas for that assignment. The report is completed outside of this notebook (delivered as a PDF). Detailed instructions for that assignment are provided in Canvas.</div>

<div class="alert alert-block alert-danger"><b><font size=4>Read this before proceeding:</font></b>
    
1. Review the list of data sets and sources of data to avoid before choosing your data. This list is provided in the instructions for the Project Proposal assignment in Canvas.<br><br>  

2. It is expected that when you are asked questions requiring typed explanations you are to use a <b><u>markdown cell</u></b> to type your answers neatly. <b><u><i>Do not provide typed answers to questions as extra comments within your code.</i></u></b> Only provide comments within your code as you normally would, i.e. as needed to explain or remind yourself what each part of the code is doing.</div>

# Project Proposal

<div class="alert alert-block alert-info">The intent of this assignment is for you to share your chosen data file(s) with your instructor and provide general information on your goals for the EDA project.</div>

<div class="alert alert-block alert-success"><b>Step 1 (2 pts)</b>: Give a brief <i><u>description</u></i> of the source(s) of your data and include a <i><u>direct link</u></i> to your data.</div>

##### Question 1

Our source of data is from IMDb (International Movie Database) which is an online database and website for information about movies, television shows, and other streaming content. It is a popular site for people to look up inforamation about ratings, cast, directors, reviews, plot, and any other related information about the content. The 4 datasets we chose to use from their database are the Titles basics, the Ratings data, the crew data, and the name basics. The Titles basics has the name of the media, when and how long it was released or aired, run time, and the genres it belongs to. The Ratings data provides the popularity rating and number of votes of the media title. The crew data provides the IDs for directors and writers. Lastly, the Name basics provides the person's name and primary professions. More information about the data sets can be found here: https://www.imdb.com/interfaces/. The specific data sets for download are found here: https://datasets.imdbws.com/.

<div class="alert alert-block alert-success"><b>Step 2 (2 pts)</b>: Briefly explain why you chose this data.</div>

##### Question 2

Understanding data is a key characteristic to starting statistical analysis. Starting off, our team discussed interests and experiences we might have in common, ultimately agreeing with true crime media. This led us to the IMDb dataset, which we could easily understand given our backgrounds and previous use of the site. Furthermore, we were able to come up with unique questions related to True Crime that we want answered through the analysis of our dataset. While there are analytical reports of IMDb data online, we did not find any analysis related to the questions we came up with. 

Out of this database, the most relevant datasets that we believe could be useful to answer our questions were title.basics, title.ratings, name.basics and title.crew. There are many other databases such as Rotten Tomatoes or Yahoo! Movies, but IMBD allows us to easily read their tsv data files through the pandas library. This is a simple file format to manipulate and extract to dive deep into coding.

<div class="alert alert-block alert-success"><b>Step 3 (1 pt)</b>: Provide a brief overview of your goals for this project.</div>

##### Question 3

We want to understand what drives the popularity of True Crime media, and media in general over time.  Specifically, we would like to be able to prove or disprove some of the following questions by the end of this project:
  * Is it true that True Crime has become more popular over the past five years? 
  * Has there been an increase in the popularity of documentaries in general?
  * Has there been an increase in the volume of documentaries in general?
  * Is there a correlation between the air date of a True Crime TV show or movie and it's popularity? 
  * Are there correlations between the popularity of a genre, specifically True Crime, and the crew?
  
Given these questions, we'll need to import data from IMDb for the title of the production itself, the crew, the names of the crew and the ratings for that production.  We'll need to download the data, clean it (there are many nulls represented by "/N") and join it in order to proceed with answering the above questions.

<div class="alert alert-block alert-success"><b>Step 4 (1 pt)</b>: Read the data into this notebook.</div>

In [105]:
##Import libraries
import sys

import pandas as pd
import urllib.request  # used to retrieve files from the internet
import numpy as np
import re
import seaborn as sns
import plotly.express as px

import plotly.offline as py
py.offline.init_notebook_mode(connected=True)

# set up notebook to display multiple output in one cell
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [106]:
def read_imdb_data(*args):
    '''
    Input a list of urls from imdb's datasets (https://datasets.imdbws.com/) and return a list of dataframes
    '''
    df_list = [] #instantiate a list
    if len(args):    # check to make sure the user input at least one item in the list
        for i in args:                     # for each url:
            filename = i.split('/', 3)[-1] # extract a filename from the url (everything after the 3rd "/" delimeter)
            urllib.request.urlretrieve(i, filename) #retrieve the file from the internet and copy it locally (https://docs.python.org/3/library/urllib.request.html)
            df_list.append(pd.read_csv(filename, compression='gzip', sep='\t', low_memory=False)) 
            # open the local file as a dataframe and append the dataframe to a list 
            # low_memory = False will ensure there no mixed types for the columns.  See https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html
    else:
        print('No URLs were passed to read_imbd_data()')
    return df_list  #returns a list of dataframes


urls = ['https://datasets.imdbws.com/title.ratings.tsv.gz',
       'https://datasets.imdbws.com/title.crew.tsv.gz',
       'https://datasets.imdbws.com/title.basics.tsv.gz',
       'https://datasets.imdbws.com/name.basics.tsv.gz']  # list of urls from imdb

df_list = read_imdb_data(*urls) # call the function with the list of urls, of any length, and save the dataframes returned

imdb_ratings, imdb_crew, imdb_title_basics, imdb_name = df_list[0], df_list[1], df_list[2], df_list[3] 
# save each dataframe independently so we can explore them

<div class="alert alert-block alert-success"><b>Step 5 (1 pt)</b>: Inspect the data using the <b>info(&nbsp;)</b>, <b>head(&nbsp;)</b>, and <b>tail(&nbsp;)</b> methods.</div>

In [107]:
# Use the info() method to determine to inspect the variable (column) names, the number of non-null values,
#       and the data types for each variable.
imdb_title_basics.info()
imdb_crew.info()
imdb_ratings.info()
imdb_name.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7634822 entries, 0 to 7634821
Data columns (total 9 columns):
 #   Column          Dtype 
---  ------          ----- 
 0   tconst          object
 1   titleType       object
 2   primaryTitle    object
 3   originalTitle   object
 4   isAdult         object
 5   startYear       object
 6   endYear         object
 7   runtimeMinutes  object
 8   genres          object
dtypes: object(9)
memory usage: 524.2+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7638150 entries, 0 to 7638149
Data columns (total 3 columns):
 #   Column     Dtype 
---  ------     ----- 
 0   tconst     object
 1   directors  object
 2   writers    object
dtypes: object(3)
memory usage: 174.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1123707 entries, 0 to 1123706
Data columns (total 3 columns):
 #   Column         Non-Null Count    Dtype  
---  ------         --------------    -----  
 0   tconst         1123707 non-null  object 
 1   averageRating 

In [108]:
# Use the head() method to inspect the first five (or more) rows of the data
imdb_title_basics.head()
imdb_crew.head()
imdb_ratings.head()
imdb_name.head()

tconst titleType            primaryTitle           originalTitle  \
0  tt0000001     short              Carmencita              Carmencita   
1  tt0000002     short  Le clown et ses chiens  Le clown et ses chiens   
2  tt0000003     short          Pauvre Pierrot          Pauvre Pierrot   
3  tt0000004     short             Un bon bock             Un bon bock   
4  tt0000005     short        Blacksmith Scene        Blacksmith Scene   

  isAdult startYear endYear runtimeMinutes                    genres  
0       0      1894      \N              1         Documentary,Short  
1       0      1892      \N              5           Animation,Short  
2       0      1892      \N              4  Animation,Comedy,Romance  
3       0      1892      \N             12           Animation,Short  
4       0      1893      \N              1              Comedy,Short

tconst  directors writers
0  tt0000001  nm0005690      \N
1  tt0000002  nm0721526      \N
2  tt0000003  nm0721526      \N
3  tt0000004  nm0721526      \N
4  tt0000005  nm0005690      \N

tconst  averageRating  numVotes
0  tt0000001            5.7      1684
1  tt0000002            6.0       207
2  tt0000003            6.5      1421
3  tt0000004            6.1       121
4  tt0000005            6.1      2221

nconst      primaryName birthYear deathYear  \
0  nm0000001     Fred Astaire      1899      1987   
1  nm0000002    Lauren Bacall      1924      2014   
2  nm0000003  Brigitte Bardot      1934        \N   
3  nm0000004     John Belushi      1949      1982   
4  nm0000005   Ingmar Bergman      1918      2007   

                     primaryProfession  \
0       soundtrack,actor,miscellaneous   
1                   actress,soundtrack   
2  actress,soundtrack,music_department   
3              actor,soundtrack,writer   
4                writer,director,actor   

                            knownForTitles  
0  tt0053137,tt0072308,tt0031983,tt0050419  
1  tt0037382,tt0038355,tt0071877,tt0117057  
2  tt0054452,tt0049189,tt0057345,tt0056404  
3  tt0078723,tt0080455,tt0077975,tt0072562  
4  tt0060827,tt0050986,tt0069467,tt0050976

In [109]:
# Use the tail() method to inspect the last five (or more) rows of the data
imdb_title_basics.tail()
imdb_crew.tail()
imdb_ratings.tail()
imdb_name.tail()

tconst  titleType               primaryTitle  \
7634817  tt9916848  tvEpisode              Episode #3.17   
7634818  tt9916850  tvEpisode              Episode #3.19   
7634819  tt9916852  tvEpisode              Episode #3.20   
7634820  tt9916856      short                   The Wind   
7634821  tt9916880  tvEpisode  Horrid Henry Knows It All   

                     originalTitle isAdult startYear endYear runtimeMinutes  \
7634817              Episode #3.17       0      2010      \N             \N   
7634818              Episode #3.19       0      2010      \N             \N   
7634819              Episode #3.20       0      2010      \N             \N   
7634820                   The Wind       0      2015      \N             27   
7634821  Horrid Henry Knows It All       0      2014      \N             10   

                          genres  
7634817      Action,Drama,Family  
7634818      Action,Drama,Family  
7634819      Action,Drama,Family  
7634820                    Short  
7634821  Animation,Comedy,Family

tconst            directors                        writers
7638145  tt9916848  nm5519454,nm5519375  nm6182221,nm1628284,nm2921377
7638146  tt9916850  nm5519375,nm5519454  nm6182221,nm1628284,nm2921377
7638147  tt9916852  nm5519375,nm5519454  nm6182221,nm1628284,nm2921377
7638148  tt9916856           nm10538645                      nm6951431
7638149  tt9916880            nm0996406            nm1482639,nm2586970

tconst  averageRating  numVotes
1123702  tt9916580            7.2         5
1123703  tt9916690            6.6         5
1123704  tt9916720            6.2        72
1123705  tt9916766            6.9        16
1123706  tt9916778            7.4        26

nconst         primaryName birthYear deathYear  \
10729849  nm9993714   Romeo del Rosario        \N        \N   
10729850  nm9993716       Essias Loberg        \N        \N   
10729851  nm9993717  Harikrishnan Rajan        \N        \N   
10729852  nm9993718         Aayush Nair        \N        \N   
10729853  nm9993719          Andre Hill        \N        \N   

                            primaryProfession knownForTitles  
10729849  animation_department,art_department      tt2455546  
10729850                                  NaN             \N  
10729851                      cinematographer      tt8736744  
10729852                      cinematographer             \N  
10729853                                  NaN             \N

<div class="alert alert-block alert-danger"><b>STOP HERE for your Project Proposal assignment. Submit your (1) original data file(s) along with (2) the completed notebook up to this point, and (3) the html file for grading and approval.</b></div>

<div class="alert alert-block alert-warning"><b>Instructor Feedback and Approval (3 pts)</b>: Your instructor will provide feedback in either the cell below this or via Canvas. You can expect one of the following point values for this portion.

<b>3 pts</b> - if your project goals and data set are both approved.<br>
<b>2 pts</b> - if your data set is approved but changes to your project goals (Step 3) are needed.<br>
<b>1 pt</b> - if your project goals are approved but your data set is not approved.<br>
<b>0 pts</b> - if neither your data set nor your project goals are approved.<br><br>
    
<i><u>As needed, follow your instructor's feeback and guidance to get on track for the remaining portions of the EDA project.</u></i>
</div>

# EDA Phase 1

<div class="alert alert-block alert-info">The overall goal of this assignment is to take all necessary steps to inspect the quality of your data and prepare the data according to your needs. For information and resources on the process of Exploratory Data Analysis (EDA), you should explore the <b><u>EDA Project Resources Module</u></b> in Canvas.

Once you’ve read through the information provided in that module and have a comfortable understanding of EDA using Python, complete steps 6 through 10 listed below to satisfy the requirements for your EDA Phase 1 assignment. **Remember to convert code cells provided to markdown cells for any typed responses to questions.**</div>

<div class="alert alert-block alert-success"><b>Step 6 (2 pts)</b>: Begin by elaborating in more detail from the previous assignment on why you chose this data?<br>
    
1. Explain what you hope to learn from this data. 
2. Do you have a hunch about what this data will reveal? (The answer to this question will be used in the Introduction section of your EDA report.)
</div>

We hope to learn about the perceived popularity of true crime media in recent years and whether this is driving an increase in true crime documentaries or if there are factors contributing. The sample questions we hope to learn: 

* Is it true that True Crime has become more popular over time?
* Has there been a change in volume of media and/or documentaries in general?
    * Could this just be the result of media content increasing overall?
    * Could this just be the result of documenatires as a genre increasing overall? 
* Has popularity of documentaries changed over time? How does this compare to other genres?
* Is there a correlation between release date and it's popularity?
* Are there correlations between popularity and the crew? For example, is a particular director scoring higher votes and ratings?

Our hunch is that there has been both an increase in volume of media overall and in the genre. However, we believe that the genre is growing faster in popularity over time than other genres.  

<div class="alert alert-block alert-success"><b>Step 7 (2 pts)</b>: Discuss the popluation and the sample:<br>
    
1. What is the population being represented by the data you’ve chosen? 
2. What is the total sample size?
</div>

The total population of our data is the total number of titles regardless ot type, if they have a rating, or any other attribute provided. It is representative of all streaming content titles on IMDB.

The sample size we will focus on for our analysis is just the overall number of unique titles for the overall media types, aka not specific tv epsisodes just the overall tv series, and those that have a signficant number of votes to showcase accurate ratings for popularity.

* Starting dataframe row count, directors: 9014736
* Ending dataframe row count, directors: 764306

* Starting dataframe row count, titles: 7634822
* Ending dataframe row count, titles: 575291

By the end of the analysis, we also added two new dataframes specifically for popularity.  Those samples are:
* Ending dataframe row count, directors popularity: 335817
* Ending dataframe row count, titles popularity: 210083

<div class="alert alert-block alert-success"><b>Step 8 (2 pts)</b>: Describe how the data was collected. For example, is this a random sample? Are sampling weights used with the data?</div>

The data that we have researched was provided by IMDB. IMDB is a website that includes many attributes of movies and tv shows. IMBD gathers data through many sources such as movie credits, interviews, press releases, etc. Due to the open nature of the website, a lot of the data is also entered through people in the industry or people visiting the site. To ensure accurate data, IMDB enforces regular quality checks. 

These datasets are not a random sample and no sample weights were used with the data. 


Sources:

https://help.imdb.com/article/imdb/general-information/where-does-the-information-on-imdb-come-from/GGD7NGF5X3ECFKNN?ref_=helpart_nav_24#


<div class="alert alert-block alert-success"><b>Step 9 (4 pts)</b>: In the Project Proposal assignment you used the info(&nbsp;) method to inspect the variables, their data types, and the number of non-null values. Using that information as a guide, provide definitions of each of your variables and their corresponding data types, i.e. a data dictionary. Also indicate which variables will be used for your purposes.</div>

After cleaning up the columns in both dataframes:

In [148]:
titles = {'Variable Name': ['tconst','primaryTitle','startYear','genre', 'nconst', 'averageRating', 'numVotes', 'titleType', 'isTrueCrime'],
        'Data Type': ['string object','string object','string object','string object', 'string object', 'float', 'integer', 'string object', 'boolean'],
        'Definition': ['title ID', 'media title', 'Release Year', 'Genre', 'Name ID', 'media rating', 'media votes', 'type of media', 'whether the title is true crime or not']
        }

title_df = pd.DataFrame(titles, columns = ['Variable Name', 'Data Type', 'Definition'])

title_df

director = {'Variable Name': ['directorName','directorID','tconst','startYear', 'isTrueCrime', 'averageRating', 'numVotes'],
        'Data Type': ['string object','string object','string object','string object', 'boolean', 'float', 'integer'],
        'Definition': ['director''s name', 'director''s id', 'title ID', 'Release Year', 'whether the title is true crime or not', 'media rating', 'media votes']
        }

director_df = pd.DataFrame(director, columns = ['Variable Name', 'Data Type', 'Definition'])

director_df

Variable Name      Data Type                              Definition
0         tconst  string object                                title ID
1   primaryTitle  string object                             media title
2      startYear  string object                            Release Year
3          genre  string object                                   Genre
4         nconst  string object                                 Name ID
5  averageRating          float                            media rating
6       numVotes        integer                             media votes
7      titleType  string object                           type of media
8    isTrueCrime        boolean  whether the title is true crime or not

Variable Name      Data Type                              Definition
0   directorName  string object                          directors name
1     directorID  string object                            directors id
2         tconst  string object                                title ID
3      startYear  string object                            Release Year
4    isTrueCrime        boolean  whether the title is true crime or not
5  averageRating          float                            media rating
6       numVotes        integer                             media votes

<div class="alert alert-block alert-success"><b>Step 10 (10 pts)</b>: For full credit in this problem you'll want to <i><u>take all necessary steps to report on the quality of the data</u></i> and <i><u>clean the data accordingly</u></i>. Some things to consider while doing this are listed below. <b>Depending on your data and goals, there may be additional steps needed than those listed here.</b>
    
1. Are there rows with missing or inconsistent values? If so, eliminate those rows from your data where appropriate.
2. Are there any outliers or duplicate rows? If so, eliminate those rows from your data where appropriate. 
At each stage of cleaning the data, state how many rows were eliminated.
3. Are you using all columns (variables) in the data? If not, are you eliminating those columns?
4. Consider some type of visual display such as a boxplot to determine any outliers. Do any outliers need removed? If so, how many were removed?

At each stage of cleaning the data, state how many rows were eliminated. <b><u><i>It is good practice to get the shape of the data before and after each step in cleaning the data and add typed explanations (in separate markdown cells) of the steps taken to clean the data.</i></u></b><br></div>
    
<div class="alert alert-block alert-info">Include the rest of your work below and insert cells where needed.</div>

#### Genre Cleanup - Step 1

* Add a column with a boolean datatype where it's True if the genre field contains both 'documentary' or 'crime', case insensitive
* Count the number of true crime documentaries in the dataset.
* Display the percent of true crime documentaries in the dataset

In [111]:
## Add a column "IsTrueCrime"
imdb_title_basics['isTrueCrime'] = (imdb_title_basics.genres.str.contains('crime', 
                flags=re.IGNORECASE, na=False)) & (imdb_title_basics.genres.str.contains('documentary', 
                flags=re.IGNORECASE, na=False))

In [112]:
TrueCrimeCount = len(imdb_title_basics[imdb_title_basics["isTrueCrime"]])  ## count the number of True items in IsTrueCrime

print(f'The total number of titles in the dataset is {len(imdb_title_basics)}.')

print(f'''The number of True Crime documentaries in the dataset is {TrueCrimeCount}, which is {round(TrueCrimeCount/len(imdb_title_basics), 3)}% of the total titles.''')

The total number of titles in the dataset is 7634822.
The number of True Crime documentaries in the dataset is 20136, which is 0.003% of the total titles.


#### Director Cleanup - Step 2

We're trying to answer: "Are there correlations between popularity and the crew? For example, is a particular director scoring higher votes and ratings?"

We considered breaking out the director column into separate fields, so there was only one director per column.  
First, we checked whether this was practical by checking for the max number of directors in the imdb_crew dataset:

In [113]:
a = 0 # instantiate a
b = 0 # instantiate b

for i in imdb_crew.index:  # loop through the crew table's index
    if len(imdb_crew.directors[i].split(",")) > a: # check to see if the current director count is greater than the last
        a, b = len(imdb_crew.directors[i].split(",")), i # save the director count and index id to a and b

print(f'The title with the most number of directors has an index id of {b} in the crew table and it has {a} directors. See below for more details.')

imdb_title_basics[imdb_title_basics.tconst == imdb_crew.tconst[b]]

The title with the most number of directors has an index id of 423900 in the crew table and it has 463 directors. See below for more details.


tconst titleType        primaryTitle    originalTitle isAdult  \
423901  tt0441074  tvSeries  Television Theater  Teatr Telewizji       0   

       startYear endYear runtimeMinutes genres  isTrueCrime  
423901      1953      \N             \N  Drama        False

Given that the max number of directors in the dataset is so large (though this could be an outlier), we don't think it's practical to separate them out into separate columns.  Instead we should build a director table consisting of:

* directorName (imdb_name.primaryName)
* directorID (imdb_crew.directors)
* tconst (imdb_crew.tconst)
* startYear (imdb_title_basics.startYear)
* isTrueCrime (imdb_title_basics.IsTrueCrime)
* averageRating (imdb_ratings.averageRating)
* numVotes (imdb_ratings.numVotes)

With this dataset, we can determine whether the average rating, weighted by the number of votes, is positively correlated with the director.  We will also be able to group by whether the titles are True Crime or not.

Next, prepare the data so that we can have a "director" dataframe with a single director per row.  For the directors column, we need the string with the director IDs to be a list.

In [114]:
#Convert the directors column, which is currently a string object, to a list and assign the result to a new column
print(type(imdb_crew.directors[0])) #confirm the type of the directors column

imdb_crew['directors_list'] = imdb_crew['directors'].str.split(',')
# split the directors string into a list and create a new column from it

imdb_crew.tail()

<class 'str'>


tconst            directors                        writers  \
7638145  tt9916848  nm5519454,nm5519375  nm6182221,nm1628284,nm2921377   
7638146  tt9916850  nm5519375,nm5519454  nm6182221,nm1628284,nm2921377   
7638147  tt9916852  nm5519375,nm5519454  nm6182221,nm1628284,nm2921377   
7638148  tt9916856           nm10538645                      nm6951431   
7638149  tt9916880            nm0996406            nm1482639,nm2586970   

                 directors_list  
7638145  [nm5519454, nm5519375]  
7638146  [nm5519375, nm5519454]  
7638147  [nm5519375, nm5519454]  
7638148            [nm10538645]  
7638149             [nm0996406]

Next "explode" the crew data so that each row is an individual director

In [149]:
directors = imdb_crew.explode('directors_list').drop(columns=['directors', 'writers'], inplace=False)
# https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.explode.html
# this will create a new dataframe where each row is an individual director.  We drop the directors and 
# writers columns

directors.rename(columns={"directors_list": "director"}, inplace = True) 
# rename the directors_list column to "director" since it's no longer a list

directors.tail()

tconst    director
7638146  tt9916850   nm5519454
7638147  tt9916852   nm5519375
7638147  tt9916852   nm5519454
7638148  tt9916856  nm10538645
7638149  tt9916880   nm0996406

Create the final directors dataframe by merging data from imbd_name, imdb_title_basics and imdb_ratings.  Print the number of shape of the dataframe before and after each merge: use a left merge (`how='left'`) so we can clean the dataframe appropriately later.

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html

In [150]:
print(f'The original dataframe has a shape of {directors.shape}')

# Merge in the director's name and drop the redundant nconst field
directors = pd.merge(directors, imdb_name[['nconst','primaryName']], left_on = 'director', right_on = 'nconst', how='left').drop(columns=['nconst'], inplace = False)
print(f'After merging in data from the imdb_name dataframe, directors has a shape of {directors.shape}')

# Merge in the IsTrueCrime field
directors = pd.merge(directors, imdb_title_basics[['tconst','isTrueCrime', 'startYear']], on = 'tconst', how='left')
print(f'After merging in data from the imdb_title_basics dataframe, directors has a shape of {directors.shape}')

# Merge in the averageRating and numVotes fields
directors = pd.merge(directors, imdb_ratings[['tconst','averageRating', 'numVotes']], on = 'tconst', how='left')
print(f'After merging in data from the imdb_ratings dataframe, directors has a shape of {directors.shape}')

directors.tail()

The original dataframe has a shape of (9018521, 2)
After merging in data from the imdb_name dataframe, directors has a shape of (9018521, 3)
After merging in data from the imdb_title_basics dataframe, directors has a shape of (9018521, 5)
After merging in data from the imdb_ratings dataframe, directors has a shape of (9018521, 7)


tconst    director       primaryName isTrueCrime startYear  \
9018516  tt9916850   nm5519454       Semih Bagci       False      2010   
9018517  tt9916852   nm5519375  Deniz Yorulmazer       False      2010   
9018518  tt9916852   nm5519454       Semih Bagci       False      2010   
9018519  tt9916856  nm10538645  Johan Planefeldt       False      2015   
9018520  tt9916880   nm0996406      Hilary Audus       False      2014   

         averageRating  numVotes  
9018516            NaN       NaN  
9018517            NaN       NaN  
9018518            NaN       NaN  
9018519            NaN       NaN  
9018520            NaN       NaN

#### Titles Cleanup - Step 3

In this step we will focus on dropping nulls, uncessary columns, and uncessary values in the columns. 

First, we will create a working copy of the title basics called media_titles. Then we will remove the nulls for the titles and genres.

In [151]:
media_titles = imdb_title_basics #Create new table for analysis
media_titles.shape  #This is the original population size

(7634822, 10)

In [152]:
null_ptitles = media_titles[media_titles.primaryTitle.isnull()]
null_ptitles

tconst  titleType primaryTitle originalTitle isAdult startYear  \
1419373  tt10790040  tvEpisode          NaN           NaN       0      2019   
3841497   tt1971246  tvEpisode          NaN           NaN       0      2011   
3934048   tt2067043  tvEpisode          NaN           NaN       0      1965   
5099580   tt4404732  tvEpisode          NaN           NaN       0      2015   
5708162   tt5773048  tvEpisode          NaN           NaN       0      2015   
6953493   tt8473688  tvEpisode          NaN           NaN       0      1987   
6985205   tt8541336  tvEpisode          NaN           NaN       0      2018   
7590764   tt9824302  tvEpisode          NaN           NaN       0      2016   

        endYear runtimeMinutes              genres  isTrueCrime  
1419373      \N             \N                  \N        False  
3841497      \N             \N           Biography        False  
3934048      \N             \N               Music        False  
5099580      \N             \N              Comedy        False  
5708162      \N             \N           Talk-Show        False  
6953493      \N             \N               Drama        False  
6985205      \N             \N  Reality-TV,Romance        False  
7590764      \N             \N         Documentary        False

In [153]:
null_otitles = media_titles[media_titles.originalTitle.isnull()]
null_otitles

tconst  titleType primaryTitle originalTitle isAdult startYear  \
1419373  tt10790040  tvEpisode          NaN           NaN       0      2019   
3841497   tt1971246  tvEpisode          NaN           NaN       0      2011   
3934048   tt2067043  tvEpisode          NaN           NaN       0      1965   
5099580   tt4404732  tvEpisode          NaN           NaN       0      2015   
5708162   tt5773048  tvEpisode          NaN           NaN       0      2015   
6953493   tt8473688  tvEpisode          NaN           NaN       0      1987   
6985205   tt8541336  tvEpisode          NaN           NaN       0      2018   
7590764   tt9824302  tvEpisode          NaN           NaN       0      2016   

        endYear runtimeMinutes              genres  isTrueCrime  
1419373      \N             \N                  \N        False  
3841497      \N             \N           Biography        False  
3934048      \N             \N               Music        False  
5099580      \N             \N              Comedy        False  
5708162      \N             \N           Talk-Show        False  
6953493      \N             \N               Drama        False  
6985205      \N             \N  Reality-TV,Romance        False  
7590764      \N             \N         Documentary        False

In [154]:
media_titles = media_titles.drop(null_ptitles.index)  
media_titles.shape 

(7634814, 10)

This only removed 8 titles. 

In [155]:
null_genres = media_titles[media_titles.genres.isnull()]
media_titles = media_titles.drop(null_genres.index)
media_titles.shape

(7634804, 10)

This ended up removing 10 titles.

In [156]:
media_titles.isnull().sum() #recheck and display the number of missing values in each column

tconst            0
titleType         0
primaryTitle      0
originalTitle     0
isAdult           0
startYear         0
endYear           0
runtimeMinutes    0
genres            0
isTrueCrime       0
dtype: int64

After reviewing the results, we noticied there are other values associated as null in our dataset like \N. We will come back to address those. 

Now, we will drop the unnecessary columns for our analysis and merge in the average ratings and number of votes. We did not see use for if the title is for adults, what the end year is, the run time minutes, or what the original title was. This will drop 4 columns, but also add in 2 from the ratings dataset.

In [157]:
media_titles = media_titles.drop(columns=['isAdult','endYear','runtimeMinutes','originalTitle'], inplace=False)
media_titles.info()                                 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7634804 entries, 0 to 7634821
Data columns (total 6 columns):
 #   Column        Dtype 
---  ------        ----- 
 0   tconst        object
 1   titleType     object
 2   primaryTitle  object
 3   startYear     object
 4   genres        object
 5   isTrueCrime   bool  
dtypes: bool(1), object(5)
memory usage: 356.8+ MB


In [158]:
media_titles.shape

(7634804, 6)

In [159]:
#merge the average ratings and votes
media_titles = pd.merge(media_titles, imdb_ratings[['tconst','averageRating', 'numVotes']], on = 'tconst', how='left')

In [160]:
media_titles.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7634804 entries, 0 to 7634803
Data columns (total 8 columns):
 #   Column         Dtype  
---  ------         -----  
 0   tconst         object 
 1   titleType      object 
 2   primaryTitle   object 
 3   startYear      object 
 4   genres         object 
 5   isTrueCrime    bool   
 6   averageRating  float64
 7   numVotes       float64
dtypes: bool(1), float64(2), object(5)
memory usage: 473.3+ MB


Next, we will remove titles that are out of scope for analysis. We will not be looking at individual TV epsiodes, audiobooks, radio series', video games, or other episodes. 

In [161]:
media_titles.titleType.value_counts()

tvEpisode       5538841
short            795153
movie            568734
video            295703
tvSeries         201862
tvMovie          130087
tvMiniSeries      35957
tvSpecial         31491
videoGame         27373
tvShort            9600
audiobook             1
episode               1
radioSeries           1
Name: titleType, dtype: int64

In [162]:
#remove tvEpisodes,videogame,radioSeries,audiobook,episode becasue they aren't relevant. For example, tvEpisode is just a subset of a tvSeries.

indextype1 = media_titles[media_titles.titleType == 'episode'].index
media_titles = media_titles.drop(indextype1)

indextype2 = media_titles[media_titles.titleType == 'audiobook'].index
media_titles = media_titles.drop(indextype2)

indextype3 = media_titles[media_titles.titleType == 'radioSeries'].index
media_titles = media_titles.drop(indextype3)

indextype4 = media_titles[media_titles.titleType == 'videoGame'].index
media_titles = media_titles.drop(indextype4)

indextype5 = media_titles[media_titles.titleType == 'tvEpisode'].index
media_titles = media_titles.drop(indextype5)


In [163]:
media_titles.titleType.value_counts()

short           795153
movie           568734
video           295703
tvSeries        201862
tvMovie         130087
tvMiniSeries     35957
tvSpecial        31491
tvShort           9600
Name: titleType, dtype: int64

In [164]:
media_titles.shape #From 7631524 to 2067987

(2068587, 8)

This dropped 5,563,537 titles from our media titles dataframe.

#### Remove Null Values - Step 4

Now we will re-address the additional null values. To account for them, we need to covert all of the `\N` values to null. We chose to drop all nulls from our directors and media_titles dataframes instead of imputing because after reviewing a sample of the data, a majority of the nulls were for title types that were not relevant to our analysis anyway. For example, TV episodes with missing directors. We also believe the most popular mainstream media is populated on IMDB which will be represented in our data analysis.


In [165]:
# first convert '/N' to nulls. Then grab the total number of null values after coversion
directors.shape # dataframe shape as a starting point
((directors.isnull())).sum() # count of nulls we start with 

directors.replace(to_replace = '\\N', value = np.NaN, inplace=True) # replace all '/N' strings with NaN
(directors.isnull()).sum() # we see the count of nulls increase

directors.dropna(inplace=True) #drop all the rows with a null value (NaN)

(directors.isnull()).sum() # we see the count of nulls go to zero
directors.shape # we see a decrease from 9010972 rows to 1266176 due to all the missing data

(9018521, 7)

tconst                 0
director               0
primaryName      3251773
isTrueCrime         3658
startYear           3658
averageRating    7581156
numVotes         7581156
dtype: int64

tconst                 0
director         3251610
primaryName      3251773
isTrueCrime         3658
startYear         863643
averageRating    7581156
numVotes         7581156
dtype: int64

tconst           0
director         0
primaryName      0
isTrueCrime      0
startYear        0
averageRating    0
numVotes         0
dtype: int64

(1265006, 7)

For our directors table, we deleted 7,744,796 rows that all had missing information. 

In [166]:
# first convert '/N' to nulls as well. Then grab the total number of null values after coversion
# titles table - conversion
media_titles.shape # dataframe shape as a starting point
((media_titles.isnull())).sum()# count of nulls we start with 

media_titles.replace(to_replace = '\\N', value = np.NaN, inplace=True) # replace all '/N' strings with NaN
(media_titles.isnull()).sum() # we see the count of nulls increase

media_titles.dropna(inplace=True) #drop all the rows with a null value (NaN)

(media_titles.isnull()).sum() # we see the count of nulls go to zero
media_titles.shape # we see a decrease from 2067987 rows to 576250 due to all the missing data

(2068587, 8)

tconst                 0
titleType              0
primaryTitle           0
startYear              0
genres                 0
isTrueCrime            0
averageRating    1475714
numVotes         1475714
dtype: int64

tconst                 0
titleType              0
primaryTitle           0
startYear         111366
genres            114298
isTrueCrime            0
averageRating    1475714
numVotes         1475714
dtype: int64

tconst           0
titleType        0
primaryTitle     0
startYear        0
genres           0
isTrueCrime      0
averageRating    0
numVotes         0
dtype: int64

(575291, 8)

We now have 576,250 titles in our media_titles data set. We dropped 1,491,737 titles due to their missing data.

#### Check for duplicates - Step 5 
Now, we will check for duplicates.

In [167]:
media_titles.head()
duplicate_media_titles = media_titles.duplicated('tconst')
duplicate_media_titles.value_counts()

tconst titleType            primaryTitle startYear  \
0  tt0000001     short              Carmencita      1894   
1  tt0000002     short  Le clown et ses chiens      1892   
2  tt0000003     short          Pauvre Pierrot      1892   
3  tt0000004     short             Un bon bock      1892   
4  tt0000005     short        Blacksmith Scene      1893   

                     genres  isTrueCrime  averageRating  numVotes  
0         Documentary,Short        False            5.7    1684.0  
1           Animation,Short        False            6.0     207.0  
2  Animation,Comedy,Romance        False            6.5    1421.0  
3           Animation,Short        False            6.1     121.0  
4              Comedy,Short        False            6.1    2221.0

False    575291
dtype: int64

In [168]:
directors.head()
duplicate_dir_tconst = directors.duplicated(subset=['tconst'])
duplicate_dir_tconst.value_counts() #313642 duplicate titles in the directors table

tconst   director           primaryName isTrueCrime startYear  \
0  tt0000001  nm0005690  William K.L. Dickson       False      1894   
1  tt0000002  nm0721526         Émile Reynaud       False      1892   
2  tt0000003  nm0721526         Émile Reynaud       False      1892   
3  tt0000004  nm0721526         Émile Reynaud       False      1892   
4  tt0000005  nm0005690  William K.L. Dickson       False      1893   

   averageRating  numVotes  
0            5.7    1684.0  
1            6.0     207.0  
2            6.5    1421.0  
3            6.1     121.0  
4            6.1    2221.0

False    951381
True     313625
dtype: int64

In [169]:
duplicate_dir_director = directors.duplicated(subset=['director'])
duplicate_dir_director.value_counts() #1019989 duplicate directors in the directors table

True     1019311
False     245695
dtype: int64

In [170]:
duplicate_dir_both = directors.duplicated()
duplicate_dir_both.value_counts() #no duplicate rows in directors table

False    1265006
dtype: int64

There were no duplicate titles in our media_titles dataset. As expected, we see duplicate titles in the directors dataset because a director can direct for multiple movies. Likewise, we see multiple titles because movies can have more than one director. However, we do not see any duplicate rows in the directors table.



#### Drop titles from the directors that aren't in the media_titles dataframe - Step 6

We're using media_titles as our primary dataframe for analysis.  The directors dataframe should only include titles that are also included in our primary dataframe so any comparisons between the two are done on the same set of titles.

In [171]:
#https://stackoverflow.com/questions/3462143/get-difference-between-two-lists
directors.shape
director_titles = list(directors.tconst)
media_titles_titles = list(media_titles.tconst)                      
excess_director_titles = list(set(director_titles) - set(media_titles_titles)) 
# returns a list of title IDs that are in director_titles, but not in media_titles_titles
len(excess_director_titles)  # tells us how many "extra" titles are in directors
directors = directors[~directors.tconst.isin(excess_director_titles)] 
# returns a list of titles that are not in excess_director_titles
directors.shape

(1265006, 7)

411786

(764306, 7)

#### Outliers and a Valid Number of Ratings based on Volatility

##### Outliers and Boxplots

Boxplots don't tell us much about the data - higher numbers of votes appear as outliers, but these are actually values that are worth keeping.  There are many more titles with few votes than there are titles with many. See charts below.

Similarly, boxplots on rating don't tell us much either - it's an ordinal scale that is useful for comparison, but we cannot tell anything concrete from the values themselves. In other words, we don't know anything about whether the differences in ratings actually correspond to anything meanful to the title they represent to actually be able to throw out outliers for the averageRating column.

In [144]:
fig = px.box(directors.sample(n=10000, random_state=1), y="numVotes", points="all", title='Directors Boxplot - Number of Votes')
fig.show()

fig = px.box(media_titles.sample(n=10000, random_state=1), y="numVotes", points="all", title='Media Titles Boxplot - Number of Votes')
fig.show()

##### Volatility and a minimum threshold for number of votes

We expect volatility for the rating of a title to be high when the number of votes is low.  We measure volatility on a sample of the data to determine an approximate minimum threshold for the number of votes to assume a reliable rating.   

Outliers on the other end of the numVotes column, i.e. movies with extremely high volumes of ratings, are not considered invalid - for this dataset we assume that the more rating there are, the better the averageRating.


Helpful articles:
1. https://www.bazaarvoice.com/blog/many-reviews-take-achieve-meaningful-average-rating/
2. https://pbpython.com/pandas-qcut-cut.html <- if we wanted to split our dataframes into equal-sized quartiles instead of sampling

In [145]:
media_titles['numVotes'].describe().apply(lambda x: format(x, 'f')) 
# describe the data (but not with scientific notation)

count     575291.000000
mean        1724.198446
std        22371.321392
min            5.000000
25%           10.000000
50%           25.000000
75%          114.000000
max      2351539.000000
Name: numVotes, dtype: object

In [146]:
media_titles_sample = media_titles.sample(n=10000, random_state=1).copy()  # take a sample of the media titles dataframe
media_titles_sample = media_titles_sample.sort_values(by=['numVotes']).set_index(['numVotes']) # sort the dataframe by the number of votes
media_titles_sample['pct_change_rating'] = media_titles_sample.averageRating.pct_change() # caluculate a rolling percent change on the same
# this will tell us for each subsequent row, how much the averageRating changed.  Higher values here indicate volatility
media_titles_sample.reset_index(inplace=True)
fig = px.line(media_titles_sample, x='numVotes', y='pct_change_rating', title='Rating Volatility by Number of Votes')
fig.show()

Zooming in on the above chart roughly indicates that around 50 votes is a good threshold to use to ensure we're able to measure popularity properly.

Therefore, we created two new dataframes specifically for popularity from the directors and media titles dataframes, dropping any titles with less than 50 votes.

In [178]:
directors_before = directors.shape
print(f'Before dropping titles with less than 9000 votes there were {directors.shape[0]} records in the directors dataframe')
directors_popularity = directors[directors['numVotes'] > 50].copy()
print(f'After dropping titles with less than 9000 votes there were {directors_popularity.shape[0]} records in the new directors_popularity dataframe')

print(f'Before dropping titles with less than 9000 votes there were {media_titles.shape[0]} records in the directors dataframe')
media_titles_popularity = media_titles[media_titles['numVotes'] > 50].copy()
print(f'After dropping titles with less than 9000 votes there were {media_titles_popularity.shape[0]} records in the new media_titles_popularity dataframe')

Before dropping titles with less than 9000 votes there were 764306 records in the directors dataframe
After dropping titles with less than 9000 votes there were 335817 records in the new directors_popularity dataframe
Before dropping titles with less than 9000 votes there were 575291 records in the directors dataframe
After dropping titles with less than 9000 votes there were 210083 records in the new media_titles_popularity dataframe


By dropping records with less than 50 votes, how many True Crime Titles are left?

In [186]:
media_titles[media_titles.isTrueCrime].shape
media_titles_popularity[media_titles_popularity.isTrueCrime].shape

(1377, 8)

(685, 8)

What are the shapes of the dataframes we end up with?

In [188]:
directors.shape
media_titles.shape
directors_popularity.shape
media_titles_popularity.shape

(764306, 7)

(575291, 8)

(335817, 7)

(210083, 8)

#### Export the Data - Step 7

In [189]:
directors.to_csv('directors.csv',index=False)
media_titles.to_csv('media_titles.csv',index=False)
directors_popularity.to_csv('directors_popularity.csv',index=False)
media_titles_popularity.to_csv('media_titles_popularity.csv',index=False)

<div class="alert alert-block alert-danger"><b>STOP HERE for your EDA Phase 1 assignment. Submit your <i><u>cleaned</u></i> data file along with the completed notebook up to this point for grading.</b></div>

# EDA Phase 2

<div class="alert alert-block alert-info">All of your work for the EDA Phase 2 assignment will begin below here. Refer to the detailed instructions and expectations for this assignment in Canvas.</div>